In [47]:
df = pd.read_csv("input002-2.txt")

In [52]:
import math
import os
import random
import re
import sys
import pandas as pd
import numpy as np

def compute_diff(data):
    ## add a new column named diff as the daily price change
    data['diff'] = data["Adj_Close"].diff()
    ## print the first 20 rows of data
    print(data.head(20))

#2    
def compute_gain_loss(data):
    pass
    data['diff'] = data["Adj_Close"].diff()
    ## add a new column named gain as the positive part of daily price change
    data['gain'] = [data['diff'][i] if data['diff'][i] > 0 else 0 for i in range(len(data["diff"]))]
    ## add a new column named loss as the negative part of daily price change
    data['loss'] = [-1*data['diff'][i] if data['diff'][i] < 0 else 0 for i in range(len(data["diff"]))]
    data['gain'][0], data['loss'][0] = data['diff'][0], data['diff'][0]
    ## print the first 20 rows of data
    print(data.head(20))

#3  
def compute_avg_gain_loss(data):
    def ma(x,n,y0):
        a = (n-1)/n
        ak = a**np.arange(len(x)-1,-1,-1)
        return np.r_[np.full(n,np.nan), y0, np.cumsum(ak*x)/ak/n + y0*np.arange(1,len(x)+1)]
    
    data['diff'] = data["Adj_Close"].diff()
    data['gain'] = [data['diff'][i] if data['diff'][i] > 0 else 0 for i in range(len(data["diff"]))]
    data['loss'] = [-1*data['diff'][i] if data['diff'][i] < 0 else 0 for i in range(len(data["diff"]))]
    data['gain'][0], data['loss'][0] = data['diff'][0], data['diff'][0]
    days = 14
    ## add a new column named avg_gain as the average gain described before
    #data['avg_gain'] = data['gain'].rolling(days).mean()
    ## add a new column named avg_loss as the average loss described before
    #data['avg_loss'] = data['loss'].rolling(days).mean()
    
    ## add a new column named avg_gain as the average gain described before
    data['avg_gain'] = ma(data.gain[days+1:].to_numpy(), days, np.nansum(data.gain.to_numpy()[:days+1])/days)
    
    ## add a new column named avg_loss as the average loss described before
    data['avg_loss'] = ma(data.loss[days+1:].to_numpy(), days, np.nansum(data.loss.to_numpy()[:days+1])/days)
    
    ## print the first 20 rows of data
    print(data.head(20))

In [53]:
compute_avg_gain_loss(df)

          Date  Adj_Close  diff  gain  loss  avg_gain  avg_loss
0   2021-01-04     133.72   NaN   NaN   NaN       NaN       NaN
1   2021-01-05     134.57  0.85  0.85  0.00       NaN       NaN
2   2021-01-06     135.99  1.42  1.42  0.00       NaN       NaN
3   2021-01-07     134.72 -1.27  0.00  1.27       NaN       NaN
4   2021-01-08     134.66 -0.06  0.00  0.06       NaN       NaN
5   2021-01-11     133.75 -0.91  0.00  0.91       NaN       NaN
6   2021-01-12     132.97 -0.78  0.00  0.78       NaN       NaN
7   2021-01-13     133.17  0.20  0.20  0.00       NaN       NaN
8   2021-01-14     131.76 -1.41  0.00  1.41       NaN       NaN
9   2021-01-15     130.77 -0.99  0.00  0.99       NaN       NaN
10  2021-01-19     129.62 -1.15  0.00  1.15       NaN       NaN
11  2021-01-20     128.00 -1.62  0.00  1.62       NaN       NaN
12  2021-01-21     127.88 -0.12  0.00  0.12       NaN       NaN
13  2021-01-22     126.89 -0.99  0.00  0.99       NaN       NaN
14  2021-01-25     129.08  2.19  2.19  0

<ipython-input-52-146619baf933>:37: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['gain'][0], data['loss'][0] = data['diff'][0], data['diff'][0]


            Adj_Close  diff  gain  loss  avg_gain  avg_loss
Date                                                       
2021-01-04     133.72   NaN   NaN   NaN       NaN       NaN
2021-01-05     134.57  0.85  0.85  0.00       NaN       NaN
2021-01-06     135.99  1.42  1.42  0.00       NaN       NaN
2021-01-07     134.72 -1.27  0.00  1.27       NaN       NaN
2021-01-08     134.66 -0.06  0.00  0.06       NaN       NaN
2021-01-11     133.75 -0.91  0.00  0.91       NaN       NaN
2021-01-12     132.97 -0.78  0.00  0.78       NaN       NaN
2021-01-13     133.17  0.20  0.20  0.00       NaN       NaN
2021-01-14     131.76 -1.41  0.00  1.41       NaN       NaN
2021-01-15     130.77 -0.99  0.00  0.99       NaN       NaN
2021-01-19     129.62 -1.15  0.00  1.15       NaN       NaN
2021-01-20     128.00 -1.62  0.00  1.62       NaN       NaN
2021-01-21     127.88 -0.12  0.00  0.12       NaN       NaN
2021-01-22     126.89 -0.99  0.00  0.99       NaN       NaN
2021-01-25     129.08  2.19  2.19  0.00  0.332857  0.664286
2021-01-26     129.91  0.83  0.83  0.00  0.368367  0.616837
2021-01-27     125.31 -4.60  0.00  4.60  0.342055  0.901348
2021-01-28     127.24  1.93  1.93  0.00  0.455480  0.836966
2021-01-29     125.14 -2.10  0.00  2.10  0.422946  0.927183
2021-02-01     125.88  0.74  0.74  0.00  0.445592  0.860956

In [33]:
df[['compute_avg_gain_loss']]

compute_avg_gain_loss
Date                   Adj_Close
2021-01-04                133.72
2021-01-05                134.57
2021-01-06                135.99
2021-01-07                134.72
...                          ...
2021-12-27                 161.1
2021-12-28                161.97
2021-12-29                163.31
2021-12-30                 161.9
2021-12-31                 162.7

[253 rows x 1 columns]

In [34]:
df1 = pd.DataFrame(df[['compute_avg_gain_loss']],index=df.index)

In [42]:
x=0
for i in df['compute_avg_gain_loss']:
    print(i)
    x+=1
    if x >10:
        break

Adj_Close
133.72
134.57
135.99
134.72
134.66
133.75
132.97
133.17
131.76
130.77


In [44]:
df['compute_avg_gain_loss'][1]

'133.72'

In [45]:
df['compute_avg_gain_loss'].index

Index(['Date', '2021-01-04', '2021-01-05', '2021-01-06', '2021-01-07',
       '2021-01-08', '2021-01-11', '2021-01-12', '2021-01-13', '2021-01-14',
       ...
       '2021-12-17', '2021-12-20', '2021-12-21', '2021-12-22', '2021-12-23',
       '2021-12-27', '2021-12-28', '2021-12-29', '2021-12-30', '2021-12-31'],
      dtype='object', length=253)